### PR TITLE
Script to remove or purge the thin-edge.io

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes

This script makes it easy to remove/purge the thin-edge.io that's installed using the Debian packages.

One can use `delete_thin_edge.sh purge/remove`
`purge` will uninstall the thin-edge packages and removes all the config files that are created.
`remove` will just uninstall the thin-edge packages

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
https://github.com/thin-edge/thin-edge.io/issues/1231

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: Cargo.lock
Comment (PradeepKiruvale) [inline Cargo.lock:1514]: These changes are not related to my PR. But mostly the Cargo.lock was not updated to use this maplit crate.

Comment (matthiasbeyer) [inline Cargo.lock:1514]: Can you split these into another commit, or preferably wait until #1321 is merged and then rebase?

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (matthiasbeyer) [inline ci/delete_thin-edge/delete_thin_edge.sh:13]:
```suggestion
    local status
    status=$(systemctl show -p ActiveState --value "$1")
```

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (matthiasbeyer) [inline ci/delete_thin-edge/delete_thin_edge.sh:28]:
```suggestion
    local status
    status=$(dpkg -s $2 | grep -w installed) && returncode=$? || returncode=$?
```

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (matthiasbeyer) [inline ci/delete_thin-edge/delete_thin_edge.sh:30]:
```suggestion
        sudo apt --assume-yes "$1" "$2"
```

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (matthiasbeyer) [inline ci/delete_thin-edge/delete_thin_edge.sh:15]:
```suggestion
        sudo systemctl stop "$1"
```

**Review by matthiasbeyer — CHANGES_REQUESTED** [review]
Maybe we should add a `shellcheck` job for these scripts.

Also, we definitively need a job to check whether `Cargo.lock` is up to date. I'm gonna file an issue for this.

File: Cargo.lock
Comment (rina23q) [reply Cargo.lock:1514]: #1321 is merged.

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (PradeepKiruvale) [reply ci/delete_thin-edge/delete_thin_edge.sh]: fixed

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (PradeepKiruvale) [reply ci/delete_thin-edge/delete_thin_edge.sh]: fixed

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (PradeepKiruvale) [reply ci/delete_thin-edge/delete_thin_edge.sh]: fixed

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (PradeepKiruvale) [reply ci/delete_thin-edge/delete_thin_edge.sh]: fixed

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (matthiasbeyer) [reply ci/delete_thin-edge/delete_thin_edge.sh]: That is not true.

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (matthiasbeyer) [reply ci/delete_thin-edge/delete_thin_edge.sh]: That is not true.

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (matthiasbeyer) [inline ci/delete_thin-edge/delete_thin_edge.sh:29]:
```suggestion
    local status
    status=$(dpkg -s "$2" | grep -w installed) && returncode=$? || returncode=$?
    if [ "$status" == "Status: install ok installed" ]; then
```

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (matthiasbeyer) [inline ci/delete_thin-edge/delete_thin_edge.sh:14]:
```suggestion
    local status
    status=$(systemctl show -p ActiveState --value "$1")
    if [ "$status" == "active" ]; then
```

**Review by matthiasbeyer — CHANGES_REQUESTED** [review]

**Review by matthiasbeyer — APPROVED** [review]
Awesome, thanks for taking the time :tada: 

Feel free to squash everything now before we merge, I think one commit should be okay for the size of this PR! :+1:

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (rina23q) [inline ci/delete_thin-edge/delete_thin_edge.sh:1]: The script name must be consistent with `get-thin-edge_io.sh`.

My proposal is `delete-thin-edge_io.sh`.

Comment (rina23q) [inline ci/delete_thin-edge/delete_thin_edge.sh:1]: Doesn't it work with `sh`? If yes, change it.
```suggestion
#!/bin/sh
```

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (rina23q) [inline ci/delete_thin-edge/delete_thin_edge.sh:25]: As far as I know, we don't need to stop apama.

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (rina23q) [inline ci/delete_thin-edge/delete_thin_edge.sh:37]: I think here one line comment will be appreciated why `tedge_mapper`, `tedge_agent`, and `tedge_watchdog` don't need to be listed. (Explain the dependency shortly after this line)

This is the dependencies around here.

1. tedge
2. tedge_mapper (dependency on `tedge`)
3. tedge_agent (dependency on `tedge_mapper`)
4. tedge_watchdog (dependency on `tedge`)

Others plugin don't have any dependencies, therefore we need to run the command explicitly.

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (rina23q) [inline ci/delete_thin-edge/delete_thin_edge.sh:74]: This is not the responsibility of this script. Remove this part.

It's the responsibility of `postrm` script of `tedge`.
https://github.com/thin-edge/thin-edge.io/blob/main/configuration/debian/tedge/postrm#L26

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (rina23q) [inline ci/delete_thin-edge/delete_thin_edge.sh:69]: Purging other software is too much. Remove this line.

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (rina23q) [reply ci/delete_thin-edge/delete_thin_edge.sh:24]: Choose either having `.service` on each command or not. Don't mix them.

**Review by rina23q — CHANGES_REQUESTED** [review]
Fix the script as commented.

Note: to close #1231, we also need another PR to add a change on [doc](https://thin-edge.github.io/thin-edge.io/html/howto-guides/002_installation.html).

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (PradeepKiruvale) [reply ci/delete_thin-edge/delete_thin_edge.sh:1]: make sense to keep them consistent. Fixed.

Comment (PradeepKiruvale) [reply ci/delete_thin-edge/delete_thin_edge.sh:1]: fixed

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (PradeepKiruvale) [reply ci/delete_thin-edge/delete_thin_edge.sh:8]: updated

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (PradeepKiruvale) [reply ci/delete_thin-edge/delete_thin_edge.sh:25]: Added `apama` because it was mentioned in the ticket, removed.

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (PradeepKiruvale) [reply ci/delete_thin-edge/delete_thin_edge.sh:24]: fixed

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (PradeepKiruvale) [reply ci/delete_thin-edge/delete_thin_edge.sh:37]: Added a comment.

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (PradeepKiruvale) [reply ci/delete_thin-edge/delete_thin_edge.sh:74]: Yes, agreed. If someone does `remove` first and then `purge` then this directory was not getting deleted, because there will be no packages. So, I thought it's good to remove it explicitly.

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (PradeepKiruvale) [reply ci/delete_thin-edge/delete_thin_edge.sh:69]: It is mentioned in the ticket, so I added it here.

**Comment by PradeepKiruvale** [comment]
> Fix the script as commented.
> 
> Note: to close #1231, we also need another PR to add a change on [doc](https://thin-edge.github.io/thin-edge.io/html/howto-guides/002_installation.html).

Sure, once the script is finalized I will update the doc. I am also confused about where to keep the script.
I feel it's better to keep the script at the top level, where the `get-thin-edge.io.sh` is placed. Any concerns?

File: ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh
Comment (albinsuresh) [inline ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh:19]: Just FYI: `systemctl` has an `is-active` subcommand as well.

File: ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh
Comment (albinsuresh) [inline ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh:25]:
```suggestion
stop_extension_services() {
```
Just to be clearer as I was wondering why you were not stopping the tedge-agent or mapper services before seeing the rest of the code.

File: ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh
Comment (albinsuresh) [inline ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh:71]: Don't need to `disconnect_from_cloud` here?

File: ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh
Comment (PradeepKiruvale) [reply ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh]: Good idea to use this.

File: ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh
Comment (PradeepKiruvale) [reply ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh]: Renamed it.

File: ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh
Comment (PradeepKiruvale) [reply ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh]: good catch, need to disconnect as well as stop the services.

File: ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh
Comment (matthiasbeyer) [inline ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh:18]:
```suggestion
    local status
    status=$(sudo systemctl is-active "$1") && returncode=$? || returncode=$?
```

And what about the returncode thing here? You're not using it anywhere...

File: ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh
Comment (matthiasbeyer) [inline ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh:32]:
```suggestion
    local status
    status=$(dpkg -s "$2" | grep -w installed) && returncode=$? || returncode=$?
```

Same question about the `returncode` variable: You're not using this anywhere...

**Review by matthiasbeyer — CHANGES_REQUESTED** [review]

File: ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh
Comment (PradeepKiruvale) [reply ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh:32]: Catching the `returncode` of the command to avoid the exiting of the script, if `returncode` is non-zero.

Comment (matthiasbeyer) [reply ci/uninstall-thin-edge_io/uninstall-thin-edge_io.sh:32]: Ah, ok.

File: uninstall-thin-edge_io.sh
Comment (rina23q) [inline uninstall-thin-edge_io.sh:75]: I think comment is placed in a wrong place, seems to be the correct place is  before `purge_packages()`.

File: ci/delete_thin-edge/delete_thin_edge.sh
Comment (rina23q) [reply ci/delete_thin-edge/delete_thin_edge.sh]: Murat just copied the commands which we use to "clean up" Raspberry Pi to fit our use-case. The script is helping user to remove `thin-edge.io`, however, better doing less than overdoing. I don't think it is good that uninstalling collectd even if the user has already installed it before the installation of thin-edge.

**Review by rina23q — APPROVED** [review]
LGTM

File: uninstall-thin-edge_io.sh
Comment (albinsuresh) [inline uninstall-thin-edge_io.sh:76]: Instead of having a dedicated `purge_thin_edge_io` function and `remove_thin_edge_io` functions with 90% duplicated code, which trickled down to `remove_packages` and `purge_packages` as well, I was wondering if it wouldn't be better to have just a single function that takes "remove" or "purge" as the argument and then just use them in the remove commands to be executed.

File: uninstall-thin-edge_io.sh
Comment (albinsuresh) [inline uninstall-thin-edge_io.sh:40]: If this function could be parametrized, to accept "purge" or "remove" as arguments, then we could avoid the code duplication between this function and `purge_packages` function.

File: uninstall-thin-edge_io.sh
Comment (PradeepKiruvale) [reply uninstall-thin-edge_io.sh:76]: Yes, can be done. But this gives a clear separation of code. The code duplication is not much, just 3/4 lines of code that's all.

File: uninstall-thin-edge_io.sh
Comment (PradeepKiruvale) [reply uninstall-thin-edge_io.sh:40]: The common factors are already abstracted in `remove_or_purge_package_if_exists`. So, I feel the existing ones are ok.

Comment (albinsuresh) [reply uninstall-thin-edge_io.sh:40]: Currently there are only a handful of packages/services, which isn't a big deal. My concern was with future additions. Whenever we add a new plugin in future, we'll end up adding it to these 2 functions. So, the refactoring suggestion was just to avoid that in future rather than the little code duplication that exists today.

File: uninstall-thin-edge_io.sh
Comment (albinsuresh) [inline uninstall-thin-edge_io.sh:59]: One of the earlier versions of this script was trying to remove `/etc/tedge` explicitly at the end of `purge`. Did you intentionally omit that here, or was it lost in refactoring? I don't think it's mandatory to have it, as purging `tedge` is supposed to delete that directory as well. But still, it was a nice safety valve.

File: uninstall-thin-edge_io.sh
Comment (albinsuresh) [inline uninstall-thin-edge_io.sh:7]: Variable names that starts with upper case looks quite unusual to my eyes, even though shell doesn't enforce any restrictions as such. Lower case names is the general convention that I've seen.

**Review by albinsuresh — APPROVED** [review]
This version definitely looks much better and brief. Keeping the package and service names in arrays was a nice change.

**Pull request merged**
